# 모르는 것
- test data의 결측치도 평균값으로 채워도 되는지??

# 내가 체크해야할 것
- int / category type 구분하기
- category column의 결측치는 'None'으로 처리해서 one-hot encoding하고
- int column의 결측치는 어떤 방식으로 채울지 고민하기

## 데이터 설명

- SalePrice - the property's sale price in dollars. This is the target variable that you're trying to predict.
- MSSubClass: The building class
- MSZoning: The general zoning classification
- LotFrontage: Linear feet of street connected to property
- LotArea: Lot size in square feet
- Street: Type of road access
- Alley: Type of alley access
- LotShape: General shape of property
- LandContour: Flatness of the property
- Utilities: Type of utilities available
- LotConfig: Lot configuration
- LandSlope: Slope of property
- Neighborhood: Physical locations within Ames city limits
- Condition1: Proximity to main road or railroad
- Condition2: Proximity to main road or railroad (if a second is present)
- BldgType: Type of dwelling
- HouseStyle: Style of dwelling
- OverallQual: Overall material and finish quality
- OverallCond: Overall condition rating
- YearBuilt: Original construction date
- YearRemodAdd: Remodel date
- RoofStyle: Type of roof
- RoofMatl: Roof material
- Exterior1st: Exterior covering on house
- Exterior2nd: Exterior covering on house (if more than one material)
- MasVnrType: Masonry veneer type
- MasVnrArea: Masonry veneer area in square feet
- ExterQual: Exterior material quality
- ExterCond: Present condition of the material on the exterior
- Foundation: Type of foundation
- BsmtQual: Height of the basement
- BsmtCond: General condition of the basement
- BsmtExposure: Walkout or garden level basement walls
- BsmtFinType1: Quality of basement finished area
- BsmtFinSF1: Type 1 finished square feet
- BsmtFinType2: Quality of second finished area (if present)
- BsmtFinSF2: Type 2 finished square feet
- BsmtUnfSF: Unfinished square feet of basement area
- TotalBsmtSF: Total square feet of basement area
- Heating: Type of heating
- HeatingQC: Heating quality and condition
- CentralAir: Central air conditioning
- Electrical: Electrical system
- 1stFlrSF: First Floor square feet
- 2ndFlrSF: Second floor square feet
- LowQualFinSF: Low quality finished square feet (all floors)
- GrLivArea: Above grade (ground) living area square feet
- BsmtFullBath: Basement full bathrooms
- BsmtHalfBath: Basement half bathrooms
- FullBath: Full bathrooms above grade
- HalfBath: Half baths above grade
- Bedroom: Number of bedrooms above basement level
- Kitchen: Number of kitchens
- KitchenQual: Kitchen quality
- TotRmsAbvGrd: Total rooms above grade (does not include bathrooms)
- Functional: Home functionality rating
- Fireplaces: Number of fireplaces
- FireplaceQu: Fireplace quality
- GarageType: Garage location
- GarageYrBlt: Year garage was built
- GarageFinish: Interior finish of the garage
- GarageCars: Size of garage in car capacity
- GarageArea: Size of garage in square feet
- GarageQual: Garage quality
- GarageCond: Garage condition
- PavedDrive: Paved driveway
- WoodDeckSF: Wood deck area in square feet
- OpenPorchSF: Open porch area in square feet
- EnclosedPorch: Enclosed porch area in square feet
- 3SsnPorch: Three season porch area in square feet
- ScreenPorch: Screen porch area in square feet
- PoolArea: Pool area in square feet
- PoolQC: Pool quality
- Fence: Fence quality
- MiscFeature: Miscellaneous feature not covered in other categories
- MiscVal: $Value of miscellaneous feature
- MoSold: Month Sold
- YrSold: Year Sold
- SaleType: Type of sale
- SaleCondition: Condition of sale

## 라이브러리

In [155]:
import numpy as np
import pandas as pd
import warnings
import matplotlib.pyplot as plt
from collections import Counter

## 설정

In [156]:
# 데이터 프레임의 모든 column을 볼 수 있도록 설정
pd.set_option('display.max_columns', None)
# pd.set_option('display.max_rows', None)

# reset
# pd.reset_option("display.max_columns")
pd.reset_option("display.max_rows")

# 소수점 둘째자리까지 출력
pd.set_option('display.float_format', lambda x: "%.2f" % x)

# warning 무시
warnings.filterwarnings(action='ignore')

## Data Loading

In [140]:
train_df = pd.read_csv("./data/house_train.csv")
test_df = pd.read_csv("./data/house_test.csv")

## Data Preprocessing

In [141]:
train_df.head(1)

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL        65.00     8450   Pave   NaN      Reg   

  LandContour Utilities LotConfig LandSlope Neighborhood Condition1  \
0         Lvl    AllPub    Inside       Gtl      CollgCr       Norm   

  Condition2 BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  \
0       Norm     1Fam     2Story            7            5       2003   

   YearRemodAdd RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  \
0          2003     Gable  CompShg     VinylSd     VinylSd    BrkFace   

   MasVnrArea ExterQual ExterCond Foundation BsmtQual BsmtCond BsmtExposure  \
0      196.00        Gd        TA      PConc       Gd       TA           No   

  BsmtFinType1  BsmtFinSF1 BsmtFinType2  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  \
0          GLQ         706          Unf           0        150          856   

  Heating HeatingQC CentralAir Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  \
0    GasA        Ex          Y      SBrkr       856       854             0   

   GrLivArea  BsmtFullBath  BsmtHalfBath  FullBath  HalfBath  BedroomAbvGr  \
0       1710             1             0         2         1             3   

   KitchenAbvGr KitchenQual  TotRmsAbvGrd Functional  Fireplaces FireplaceQu  \
0             1          Gd             8        Typ           0         NaN   

  GarageType  GarageYrBlt GarageFinish  GarageCars  GarageArea GarageQual  \
0     Attchd      2003.00          RFn           2         548         TA   

  GarageCond PavedDrive  WoodDeckSF  OpenPorchSF  EnclosedPorch  3SsnPorch  \
0         TA          Y           0           61              0          0   

   ScreenPorch  PoolArea PoolQC Fence MiscFeature  MiscVal  MoSold  YrSold  \
0            0         0    NaN   NaN         NaN        0       2    2008   

  SaleType SaleCondition  SalePrice  
0       WD        Normal     208500

In [142]:
test_df.head(1)

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0  1461          20       RH        80.00    11622   Pave   NaN      Reg   

  LandContour Utilities LotConfig LandSlope Neighborhood Condition1  \
0         Lvl    AllPub    Inside       Gtl        NAmes      Feedr   

  Condition2 BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  \
0       Norm     1Fam     1Story            5            6       1961   

   YearRemodAdd RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  \
0          1961     Gable  CompShg     VinylSd     VinylSd       None   

   MasVnrArea ExterQual ExterCond Foundation BsmtQual BsmtCond BsmtExposure  \
0        0.00        TA        TA     CBlock       TA       TA           No   

  BsmtFinType1  BsmtFinSF1 BsmtFinType2  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  \
0          Rec      468.00          LwQ      144.00     270.00       882.00   

  Heating HeatingQC CentralAir Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  \
0    GasA        TA          Y      SBrkr       896         0             0   

   GrLivArea  BsmtFullBath  BsmtHalfBath  FullBath  HalfBath  BedroomAbvGr  \
0        896          0.00          0.00         1         0             2   

   KitchenAbvGr KitchenQual  TotRmsAbvGrd Functional  Fireplaces FireplaceQu  \
0             1          TA             5        Typ           0         NaN   

  GarageType  GarageYrBlt GarageFinish  GarageCars  GarageArea GarageQual  \
0     Attchd      1961.00          Unf        1.00      730.00         TA   

  GarageCond PavedDrive  WoodDeckSF  OpenPorchSF  EnclosedPorch  3SsnPorch  \
0         TA          Y         140            0              0          0   

   ScreenPorch  PoolArea PoolQC  Fence MiscFeature  MiscVal  MoSold  YrSold  \
0          120         0    NaN  MnPrv         NaN        0       6    2010   

  SaleType SaleCondition  
0       WD        Normal

- set index 'id'

In [143]:
train_df.set_index("Id", inplace=True)
test_df.set_index("Id", inplace=True)

- train / test index 저장

In [144]:
train_index = train_df.index
test_index = test_df.index
train_index, test_index

(Int64Index([   1,    2,    3,    4,    5,    6,    7,    8,    9,   10,
             ...
             1451, 1452, 1453, 1454, 1455, 1456, 1457, 1458, 1459, 1460],
            dtype='int64', name='Id', length=1460),
 Int64Index([1461, 1462, 1463, 1464, 1465, 1466, 1467, 1468, 1469, 1470,
             ...
             2910, 2911, 2912, 2913, 2914, 2915, 2916, 2917, 2918, 2919],
            dtype='int64', name='Id', length=1459))

- y_train_df 따로 저장

In [145]:
y_train_df = train_df.pop("SalePrice")

In [146]:
y_train_df.head(3)

Id
1    208500
2    181500
3    223500
Name: SalePrice, dtype: int64

- 결측치 처리

In [148]:
train_missval = train_df.isnull().sum() / len(train_df) * 100
train_missval = train_missval[train_missval.values > 0]
train_missval

LotFrontage    17.74
Alley          93.77
MasVnrType      0.55
MasVnrArea      0.55
BsmtQual        2.53
BsmtCond        2.53
BsmtExposure    2.60
BsmtFinType1    2.53
BsmtFinType2    2.60
Electrical      0.07
FireplaceQu    47.26
GarageType      5.55
GarageYrBlt     5.55
GarageFinish    5.55
GarageQual      5.55
GarageCond      5.55
PoolQC         99.52
Fence          80.75
MiscFeature    96.30
dtype: float64

In [147]:
test_missval = test_df.isnull().sum() / len(test_df) * 100
test_missval = test_missval[test_missval > 0]
test_missval

MSZoning        0.27
LotFrontage    15.56
Alley          92.67
Utilities       0.14
Exterior1st     0.07
Exterior2nd     0.07
MasVnrType      1.10
MasVnrArea      1.03
BsmtQual        3.02
BsmtCond        3.08
BsmtExposure    3.02
BsmtFinType1    2.88
BsmtFinSF1      0.07
BsmtFinType2    2.88
BsmtFinSF2      0.07
BsmtUnfSF       0.07
TotalBsmtSF     0.07
BsmtFullBath    0.14
BsmtHalfBath    0.14
KitchenQual     0.07
Functional      0.14
FireplaceQu    50.03
GarageType      5.21
GarageYrBlt     5.35
GarageFinish    5.35
GarageCars      0.07
GarageArea      0.07
GarageQual      5.35
GarageCond      5.35
PoolQC         99.79
Fence          80.12
MiscFeature    96.50
SaleType        0.07
dtype: float64

- Decision 1 - drop Alley, PoolQC, Fence, MiscFeature
    - 결측치가 너무 많아서 삭제
    - Alley : 골목 접근 유형 (결측치 93.77%)
    - PoolQC : 수영장 품질 (결측치 99.52%)
    - Fence : 울타리 품질 (결측치 80.75%)
    - MiscFeature : 다른 카테고리에서 다루지 않는 기타 기능 -  (결측치 96.30%)

In [149]:
del train_df["Alley"]
del train_df["PoolQC"]
del train_df["Fence"]
del train_df["MiscFeature"]
del test_df["Alley"]
del test_df["PoolQC"]
del test_df["Fence"]
del test_df["MiscFeature"]

- 데이터 모판 생성 (y 제외, train + test)

In [160]:
all_df = pd.concat([train_df, test_df], axis = 0)
all_df.tail()

MSSubClass MSZoning  LotFrontage  LotArea Street LotShape LandContour  \
Id                                                                            
2915         160       RM        21.00     1936   Pave      Reg         Lvl   
2916         160       RM        21.00     1894   Pave      Reg         Lvl   
2917          20       RL       160.00    20000   Pave      Reg         Lvl   
2918          85       RL        62.00    10441   Pave      Reg         Lvl   
2919          60       RL        74.00     9627   Pave      Reg         Lvl   

     Utilities LotConfig LandSlope Neighborhood Condition1 Condition2  \
Id                                                                      
2915    AllPub    Inside       Gtl      MeadowV       Norm       Norm   
2916    AllPub    Inside       Gtl      MeadowV       Norm       Norm   
2917    AllPub    Inside       Gtl      Mitchel       Norm       Norm   
2918    AllPub    Inside       Gtl      Mitchel       Norm       Norm   
2919    AllPub    Inside       Mod      Mitchel       Norm       Norm   

     BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  YearRemodAdd  \
Id                                                                            
2915    Twnhs     2Story            4            7       1970          1970   
2916   TwnhsE     2Story            4            5       1970          1970   
2917     1Fam     1Story            5            7       1960          1996   
2918     1Fam     SFoyer            5            5       1992          1992   
2919     1Fam     2Story            7            5       1993          1994   

     RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  MasVnrArea  \
Id                                                                       
2915     Gable  CompShg     CemntBd     CmentBd       None        0.00   
2916     Gable  CompShg     CemntBd     CmentBd       None        0.00   
2917     Gable  CompShg     VinylSd     VinylSd       None        0.00   
2918     Gable  CompShg     HdBoard     Wd Shng       None        0.00   
2919     Gable  CompShg     HdBoard     HdBoard    BrkFace       94.00   

     ExterQual ExterCond Foundation BsmtQual BsmtCond BsmtExposure  \
Id                                                                   
2915        TA        TA     CBlock       TA       TA           No   
2916        TA        TA     CBlock       TA       TA           No   
2917        TA        TA     CBlock       TA       TA           No   
2918        TA        TA      PConc       Gd       TA           Av   
2919        TA        TA      PConc       Gd       TA           Av   

     BsmtFinType1  BsmtFinSF1 BsmtFinType2  BsmtFinSF2  BsmtUnfSF  \
Id                                                                  
2915          Unf        0.00          Unf        0.00     546.00   
2916          Rec      252.00          Unf        0.00     294.00   
2917          ALQ     1224.00          Unf        0.00       0.00   
2918          GLQ      337.00          Unf        0.00     575.00   
2919          LwQ      758.00          Unf        0.00     238.00   

      TotalBsmtSF Heating HeatingQC CentralAir Electrical  1stFlrSF  2ndFlrSF  \
Id                                                                              
2915       546.00    GasA        Gd          Y      SBrkr       546       546   
2916       546.00    GasA        TA          Y      SBrkr       546       546   
2917      1224.00    GasA        Ex          Y      SBrkr      1224         0   
2918       912.00    GasA        TA          Y      SBrkr       970         0   
2919       996.00    GasA        Ex          Y      SBrkr       996      1004   

      LowQualFinSF  GrLivArea  BsmtFullBath  BsmtHalfBath  FullBath  HalfBath  \
Id                                                                              
2915             0       1092          0.00          0.00         1         1   
2916             0       1092          0.00          0.00         1         1   


In [166]:
all_missval = all_df.isnull().sum() / len(all_df) * 100
all_missval[all_missval.values > 0]

MSZoning        0.14
LotFrontage    16.65
Utilities       0.07
Exterior1st     0.03
Exterior2nd     0.03
MasVnrType      0.82
MasVnrArea      0.79
BsmtQual        2.77
BsmtCond        2.81
BsmtExposure    2.81
BsmtFinType1    2.71
BsmtFinSF1      0.03
BsmtFinType2    2.74
BsmtFinSF2      0.03
BsmtUnfSF       0.03
TotalBsmtSF     0.03
Electrical      0.03
BsmtFullBath    0.07
BsmtHalfBath    0.07
KitchenQual     0.03
Functional      0.07
FireplaceQu    48.65
GarageType      5.38
GarageYrBlt     5.45
GarageFinish    5.45
GarageCars      0.03
GarageArea      0.03
GarageQual      5.45
GarageCond      5.45
SaleType        0.03
dtype: float64